QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java

```diff
@@ -199,35 +207,79 @@ public String getData(HttpServletRequest request, HttpServletResponse response,
     @ResponseBody
     @RequestMapping(value = "/query", method = RequestMethod.POST)
     public String query(HttpServletRequest request, HttpServletResponse response,
-            @RequestHeader("Accept") String accept, @RequestBody String query, @RequestParam("class") String clas)
-                    throws JSONException {
+            @RequestHeader("Accept") String accept, @RequestBody String query, @RequestParam("class") String clas) {
 
         query = query.trim();
+
+        response.setContentType(accept);
+
         if (query.isEmpty()) {
             response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
-            return new Error("Please provide the query to execute the request").toString(accept);
+            return new Message("Please provide the query to execute the request").toString(accept);
         }
 
-        Class<?> clazz = classMap.get(clas);
-        if (clazz == null) {
-            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
-            return new Error("Please provide a valid Class to execute the request").toString(accept);
+        Map<String, String> responseParams = new HashMap<>();
+        try {
+            int queryID = query.hashCode();
+            responseParams.put("pollurl", "/query/" + queryID);
+            ExecutionStatus queryStatus = asyncQueryManager.getQueryStatus(query.hashCode());
+
+            // if the query is running or completed notify the client
+            if (queryStatus == ExecutionStatus.RUNNING || queryStatus == ExecutionStatus.COMPLETED) {
+                response.setStatus(HttpServletResponse.SC_OK);
+                responseParams.put("queryStatus", queryStatus.name());
+                return new Message(responseParams).toString(accept);
+            }
+
+            Class<?> clazz = classMap.get(clas);
+            if (clazz == null) {
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                return new Message("Please provide a valid Class to execute the request").toString(accept);
+            }
+
+            if (accept != null && accept.equals(RETURN_JSON)) {
+                repositorymanager.executeQueryAsync(query, clazz, JSON, queryID);
+            } else {
+                repositorymanager.executeQueryAsync(query, clazz, XML, queryID);
```

Comment:
the format of the results needs to be decided when requesting the results, not when creating them.